codec/codec_test.go

```diff
@@ -1290,6 +1290,100 @@ func doTestMsgpackRpcSpecPythonClientToGoSvc(t *testing.T) {
 		fmt.Sprintf("%#v\n%#v\n", []string{"A1", "B2", "C3"}, TestRpcABC{"Aa", "Bb", "Cc"}), "cmdout=")
 }
 
+type UBase struct {
+	S string
+	B bool
+	I int
+}
+
+type U1 struct {
+	UBase
+
+	unknownFields map[string][]byte
+}
+
+func (u1 *U1) CodecOnUnknownField(fieldName string, encodedValue []byte) {
+	if u1.unknownFields == nil {
+		u1.unknownFields = make(map[string][]byte)
+	}
+	u1.unknownFields[fieldName] = encodedValue
+}
+
+func (u1 *U1) GetUnknownFields() map[string][]byte {
+	fmt.Printf("U1.GetUnknownFields called, returning %+v\n", u1.unknownFields)
+	return u1.unknownFields
+}
+
+var _ UnknownFieldsHandler = (*U1)(nil)
+
+type U2 struct {
+	UBase
+
+	S2 string
+	B2 bool
+	I2 int
+
+	unknownFields map[string][]byte
+}
+
+var _ UnknownFieldsHandler = (*U2)(nil)
+
+func (u2 *U2) CodecOnUnknownField(fieldName string, encodedValue []byte) {
+	panic("Unexpected")
+}
+
+func (u2 *U2) GetUnknownFields() map[string][]byte {
+	fmt.Printf("U2.GetUnknownFields called, returning %+v\n", u2.unknownFields)
+	return u2.unknownFields
+}
+
+func doTestEncUnknownFields(t *testing.T, h Handle) {
+	u2 := U2{UBase{"t1", true, 5}, "t2", false, 3, nil}
+
+	var bs []byte
+	var err error
+	err = NewEncoderBytes(&bs, h).Encode(&u2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var u1 U1
+	err = NewDecoderBytes(bs, h).Decode(&u1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(u1.UBase, u2.UBase) {
+		t.Fatalf("u1.UBase=%+v != u2.UBase=%+v", u1.UBase, u2.UBase)
+	}
+
+	expectedUnknownFields := map[string][]byte{
+		"B2": []byte{194},
+		"I2": []byte{3},
+		"S2": []byte{162, 116, 50},
+	}
+
+	if !reflect.DeepEqual(expectedUnknownFields, u1.unknownFields) {
+		t.Fatalf("expectedUnknownFields=%+v != u1.unknownFields=%+v",
+			expectedUnknownFields, u1.unknownFields)
+	}
+
+	err = NewEncoderBytes(&bs, h).Encode(&u1)
```

Comment:
Maybe use a fresh `[]byte` here, just to prove that `NewEncoderBytes` is correctly clearing out the old buffer?